# Project Description

**Purpose** <br>
This project consists of analyzing several datasets to help in the research efforts of a new Chicago ride share company. The analysis will provide insights on trends in taxi service including popular service providers and neighborhoods. 

**Datasets** <br>
The datasets used are from data extracted from a website with weather condition information as well as SQL database queriers. 

The first dataset includes taxi companies names and the number of rides they completed on November 15-16, 2017. The secodn dataset used contains drop-off locations in Chicago neighborhoods where rides ended and the average number of rides that ended in each neighborhood in November 2017.

**Analysis** <br>
The analysis involves creating simple visualizations to identify the most frequently used Taxi services in Chicago and the top 10 neighborhoods in terms of drop-offs. These visualizations will compare different taxi service providers using bar charts. 

The final part of this project includes hypothesis testing using the data from result_07.
This dataset includes information about rides from the Loop to O'Hare International Airport, including pickup date and time, weather conditions at the time of pickup, and ride duration in seconds. 

The hypothesis to be tested is: "The average duration of rides from the Loop to O'Hare International Airport changes on rainy Saturdays."

Conducting this test will determine whether there is a significant difference between average ride times to the Airport on Saturdays depending on the weather condition. This may be helpful for planning the amount of rides or drivers set out on Saturdays by the taxi service in order to provide the best service and expectation

# Step 4: Exploratory Data Analysis

## Import Data 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
from scipy import stats

## Read Data

In [2]:
# Read the CSV files 
company_trips = pd.read_csv('moved_project_sql_result_01.csv')
average_trip_location = pd.read_csv('moved_project_sql_result_04.csv')


In [3]:
#Viewing 10 representative rows of data 
company_trips.sample(10)


company_name  trips_amount
17                  Chicago Medallion Management          1955
14          Patriot Taxi Dba Peace Taxi Associat          2235
42                   1085 - 72312 N and W Cab Co            29
43                2809 - 95474 C & D Cab Co Inc.            29
52  3721 - Santamaria Express, Alvaro Santamaria            14
57                                   Metro Group            11
6                                   City Service          8448
41                          Chicago Star Taxicab            29
11                                    Globe Taxi          4383
50                  2823 - 73307 Lee Express Inc            15

Data types appear correct but some entries seem to have numeric values at the beginning, such as "1085 - 72312 N and W Cab Co". This could be an id from a different table that was mergered incorrectly but it is useless as is, so we will just delete these values from all company_name values. 

In [4]:
# Using regular expression to remove numeric values from company names for clarity
company_trips['company_name'] = company_trips['company_name'].str.replace(r'^\d+\s*-\s*|\d+\s*', '')
company_trips['company_name'].sample(10)

29                         303 Taxi
60          2733 - 74600 Benny Jona
63      3556 - 36214 RC Andrews Cab
33                 Metro Jet Taxi A
3                        Yellow Cab
39      0118 - 42111 Godfrey S.Awir
8         Star North Management LLC
28                      5 Star Taxi
47    4615 - 83503 Tyrone Henderson
6                      City Service
Name: company_name, dtype: object

In [5]:
#Viewing 10 representative rows of data 
average_trip_location.sample(10)


dropoff_location_name  average_trips
83           Avalon Park       6.966667
58             Englewood      25.433333
39        Jefferson Park      74.333333
91              Burnside       2.333333
54        North Lawndale      32.166667
27              Avondale     208.233333
67        Auburn Gresham      15.166667
11          East Village    1212.066667
82         South Deering       7.500000
10          Lincoln Park    1246.533333

## Ensure Correct Data Types 

In [6]:
# Viewing basic data information 
company_trips.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   company_name  64 non-null     object
 1   trips_amount  64 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.1+ KB


Data types appear correct but further investigation is necessary to ensure there are no abnmormalities.

In [7]:
# Viewing basic data information 
average_trip_location.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94 entries, 0 to 93
Data columns (total 2 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   dropoff_location_name  94 non-null     object 
 1   average_trips          94 non-null     float64
dtypes: float64(1), object(1)
memory usage: 1.6+ KB


Data types appear correct but further investigation is necessary to ensure there are no abnmormalities.

In [8]:
company_trips.describe()


trips_amount
count     64.000000
mean    2145.484375
std     3812.310186
min        2.000000
25%       20.750000
50%      178.500000
75%     2106.500000
max    19558.000000

All values seem normal, showing the taxi companies with the least amount of trips completed at the min() value of 2 trips and the company with the most at the max() value of 19,558 trips completed. 

In [9]:
average_trip_location.describe()


average_trips
count      94.000000
mean      599.953728
std      1714.591098
min         1.800000
25%        14.266667
50%        52.016667
75%       298.858333
max     10727.466667

All values seem normal, showing the least and most popular drop off locations at the min() value of 1.8 trips on average and max() value of 10,727 trips on average at the most popular destination.  

The only issue is that the float is not the easiest to read as is. I debated removing changing it to integers but will round the decimals up instead. 

In [10]:
# Rounding the decimal points for readability 
average_trip_location['average_trips'] = average_trip_location['average_trips'].round(2)
average_trip_location


dropoff_location_name  average_trips
0                   Loop       10727.47
1            River North        9523.67
2          Streeterville        6664.67
3              West Loop        5163.67
4                 O'Hare        2546.90
..                   ...            ...
89       Mount Greenwood           3.14
90             Hegewisch           3.12
91              Burnside           2.33
92             East Side           1.96
93             Riverdale           1.80

[94 rows x 2 columns]

## Data Visualizations 


**Total Trips Completed by Each Taxi Company**


In [11]:
# Creates bar graph for total trips by company 
fig = px.bar(company_trips, x='trips_amount', y='company_name', 
             title='Total Taxi Trips by Company',
             category_orders={"company_name": company_trips['company_name'][::-1]})
fig.update_layout(width=800, height=600, xaxis_title='Total Trips Completed', yaxis_title='Company Name', xaxis_tickangle=0)
fig.show()


The top taxi service, Flash Cab has a great lead on total trips compeleted, showing that it is clearly very seasoned in the field and has a large fleet and established presence compared to all other services in the area. 

There are only a very few dominant players in the Chicago taxi business with only the top 12 companies appear to have completed a total of at least 100 trips, which makes it clear that there are only a handful of extremely popular services which most people use in the Chicago area. 



### Top Taxi Services 

In [12]:
# Top 10 taxi services by total trips completed 
top_10_company = company_trips.sort_values(by='trips_amount', ascending=False)
top_10_company = top_10_company .head(10)
top_10_company 

company_name  trips_amount
0                          Flash Cab         19558
1          Taxi Affiliation Services         11422
2                   Medallion Leasin         10367
3                         Yellow Cab          9888
4    Taxi Affiliation Service Yellow          9299
5          Chicago Carriage Cab Corp          9181
6                       City Service          8448
7                           Sun Taxi          7701
8          Star North Management LLC          7455
9  Blue Ribbon Taxi Association Inc.          5953

<br> 

**Top 10 Taxi Companies by Total of Trips Completed** 

In [13]:
# Creates bar graph for top 10 companies
fig = px.bar(top_10_company , x='trips_amount', y='company_name', 
             title='Top 10 Taxi Companies',
             category_orders={"company_name": company_trips['company_name'][::-1]})
fig.update_layout(xaxis_title='Total Trips Completed', yaxis_title='Company Name', xaxis_tickangle=0)
fig.show()

Again, it appears clear that 'Flash Cab' is in it's own league, while the rest of the taxi services are close competition to eachother. 

### Top 10 Neighborhood Destinations



**Top 10 Neighborhoods by Number of Dropoffs**
<br> 


In [14]:
# Top 10 drop-off locations by neighborhood 
top_10_dropoff = average_trip_location.sort_values(by='average_trips', ascending=False)
top_10_dropoff= top_10_dropoff.head(10)
top_10_dropoff

dropoff_location_name  average_trips
0                  Loop       10727.47
1           River North        9523.67
2         Streeterville        6664.67
3             West Loop        5163.67
4                O'Hare        2546.90
5             Lake View        2420.97
6            Grant Park        2068.53
7         Museum Campus        1510.00
8            Gold Coast        1364.23
9    Sheffield & DePaul        1259.77

In [15]:
# Creates bar graph for top 10 drop offs
fig = px.bar(top_10_dropoff, x='average_trips', y='dropoff_location_name', 
             title='Top 10 Neighborhood Destinations',
             category_orders={"dropoff_location_name": top_10_dropoff['dropoff_location_name'][::-1]})
fig.update_layout(width=800, height=600, xaxis_title='Average Trips Completed', yaxis_title='Drop Off Locations', xaxis_tickangle=0)
fig.show()


**Top Neighborhood Destinations** 

#1  Loop

#2 River North

#3 Streeterville

#4 West Loop

<br>

After the top 4 destinations listed above, there is a significant drop in trip totals for all following neighborhoods. This means that the top 4 are especially popular. It was suprising to see that O'Hare- the airport destination is not higher on the list, but this might be because certain commuter locations may be more popular for everyday taxi rides. 

# Step 5: Testing hypotheses (Python)

The last dataset used will be the result of the final SQL query and contains data on rides from the Loop to O'Hare International Airport. 

<br>

The table's field values are:

start_ts- This is the pickup date and time

weather_conditions- The weather conditions at the moment the ride started

duration_seconds- The ride duration in seconds

## Importing and Data Cleaning  

In [16]:
# Imports data 
loop_to_airport = pd.read_csv('moved_project_sql_result_07.csv')
loop_to_airport.sample(10)

start_ts weather_conditions  duration_seconds
71    2017-11-11 08:00:00               Good            1140.0
146   2017-11-11 14:00:00               Good            2340.0
413   2017-11-11 04:00:00               Good            1219.0
719   2017-11-18 12:00:00                Bad            3600.0
602   2017-11-11 16:00:00               Good            2220.0
939   2017-11-11 07:00:00               Good            1425.0
1062  2017-11-11 06:00:00               Good            1500.0
956   2017-11-04 10:00:00               Good            2460.0
841   2017-11-04 12:00:00               Good            2520.0
588   2017-11-04 18:00:00                Bad            2580.0

From this sample, it appears that the column 'start_ts' only shows the hour of pickup, although the current format suggests that it has the information up to seconds. The column may not be set to date time. The name of the column could also be changed for clarity. 

The 'weather_conditions' column looks great so far. 

The 'duration_seconds'column could be simplified to just integers instead of float.

In [17]:
# Check dataframe information 
loop_to_airport.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1068 entries, 0 to 1067
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   start_ts            1068 non-null   object 
 1   weather_conditions  1068 non-null   object 
 2   duration_seconds    1068 non-null   float64
dtypes: float64(1), object(2)
memory usage: 25.2+ KB


In [18]:
# Making changes to dataframe columns
loop_to_airport['duration_seconds'] = loop_to_airport['duration_seconds'].astype(int)

# Renaming and changing data type to show clear pickup time
loop_to_airport = loop_to_airport.rename(columns={'start_ts': 'pickup_date_time'})
loop_to_airport['pickup_date_time'] = pd.to_datetime(loop_to_airport['pickup_date_time']) 

In [19]:
# Confirming changes 
loop_to_airport.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1068 entries, 0 to 1067
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   pickup_date_time    1068 non-null   datetime64[ns]
 1   weather_conditions  1068 non-null   object        
 2   duration_seconds    1068 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 25.2+ KB


### Filtering Data by Day of Week = Saturday 

In this section we will only work with data from the Loop to Airport Saturdays in November 2017. 
<br>
Those Saturdays are Novemeber: the 4th, 11, 18 and 25.

In [20]:
# Filters for Saturdays in November 2017
saturday_rides = loop_to_airport[loop_to_airport['pickup_date_time'].dt.dayofweek == 5]
saturday_rides

pickup_date_time weather_conditions  duration_seconds
0    2017-11-25 16:00:00               Good              2410
1    2017-11-25 14:00:00               Good              1920
2    2017-11-25 12:00:00               Good              1543
3    2017-11-04 10:00:00               Good              2512
4    2017-11-11 07:00:00               Good              1440
...                  ...                ...               ...
1063 2017-11-25 11:00:00               Good                 0
1064 2017-11-11 10:00:00               Good              1318
1065 2017-11-11 13:00:00               Good              2100
1066 2017-11-11 08:00:00               Good              1380
1067 2017-11-04 16:00:00                Bad              2834

[1068 rows x 3 columns]

In [21]:
# Checking that the Saturday dataframe is clear 
saturday_rides.sample(10)

pickup_date_time weather_conditions  duration_seconds
772 2017-11-18 08:00:00                Bad              1260
521 2017-11-04 15:00:00               Good              3351
781 2017-11-11 12:00:00               Good              1993
20  2017-11-11 13:00:00               Good              2250
513 2017-11-04 14:00:00               Good              3286
620 2017-11-11 10:00:00               Good              1200
55  2017-11-25 14:00:00               Good              1620
65  2017-11-04 18:00:00                Bad              2363
660 2017-11-11 07:00:00               Good              1355
653 2017-11-11 10:00:00               Good              1516

### Preparing Data for Tests 

In [22]:
# Splitting data up into bad and good weather by  weather_condition and duration_seconds 
no_rain_saturdays = saturday_rides[saturday_rides['weather_conditions'] == 'Good']['duration_seconds']

rain_saturdays = saturday_rides[saturday_rides['weather_conditions'] == 'Bad']['duration_seconds']

Identifying the variance first to determine the alpha level for testing:

In [23]:
# Checking Indivudal variances
no_rain_var = no_rain_saturdays.var()
rain_var = rain_saturdays.var()

print(f'The variance of Saturdays with no rain is {no_rain_var}')
print(f'The variance of rainy Saturdays is: {rain_var}')
# Levene's test for variance comparison
levene_test = stats.levene(rain_saturdays, no_rain_saturdays)
levene_test

The variance of Saturdays with no rain is 576382.0096895089
The variance of rainy Saturdays is: 520294.0860024826


LeveneResult(statistic=0.38853489683656073, pvalue=0.5332038671974493)

Although the variance is higher for Saturdays with no rain, the difference between the two variances doesn't seem significant. 

The p-value of the Levene test is larger than the typical significance level of 0.05. Therefore, we can assume equal variances for now during testing.

In [24]:
# Double checking 
# Finding the mean and standard deviation of the dataset to compare with the variance 
no_rain_saturdays.describe()

count     888.000000
mean     1999.675676
std       759.198268
min         0.000000
25%      1389.750000
50%      1800.000000
75%      2460.000000
max      7440.000000
Name: duration_seconds, dtype: float64

In [25]:
# Finding the mean and standard deviation of the dataset to compare with the variance 
rain_saturdays.describe()

count     180.000000
mean     2427.205556
std       721.314138
min       480.000000
25%      1962.000000
50%      2540.000000
75%      2928.000000
max      4980.000000
Name: duration_seconds, dtype: float64

#### Comparing mean, standard deviation (std), and variance 
Comparing the mean, standard deviation (std), and variance for rides from the Loop to O'Hare International Airport is an extra step we can take to determine significance level for testing. 

**Saturdays without rain:**<br>
Mean: 1999.68 seconds (33.32 minutes)<br>
Standard deviation: 759.20 seconds<br>
Variance: 576382.01 seconds<br>

On Saturdays without rain, the mean duration of rides is approximately 1999.68 seconds, with a standard deviation of 759.20 seconds. This large variance of 576382.01 seconds² suggests considerable variability in ride durations around the mean.


**Rainy Saturdays:**
Mean: 2427.21 seconds (40.45 minutes) <br>
Standard deviation: 721.31 seconds<br>
Variance: 520294.09 seconds<br>

In contrast, rainy Saturdays show slightly longer mean ride durations, approximately 2427.21 seconds. Rainy Saturdays have a standard deviation of 721.31 seconds, slightly smaller than Saturdays without rain, suggesting more consistency. The variance for rainy Saturdays is also slightly lower at 520294.09 seconds² compared to Saturdays without rain.

When comparing the average durations in minutes, it's clear that Rainy Saturdays have longer rides, lasting approximately 40.45 minutes compared to 33.32 minutes on Non-Rainy Saturdays. This difference of approximately 13 extra minutes can feel significantly longer to passengers, especially considering that the average ride should take approximately 34.52 minutes.

Overall, there seems to be great variability between rides regardless of the weather on Saturdays, likely due to factors such as traffic conditions, construction, or peak hours for ride shares. Both non-rainy Saturdays and rainy Saturdays have variances much larger than the entire dataset mean of 2071.73 seconds, indicating a wide range of ride durations. Conducting a t-test will provide a better understanding of whether there is a significant difference between the two variables, helping determine if rain truly affects ride durations significantly.



### Conducting Hypothesis Testing

Since we're comparing the means of two independent samples (rides duration on rainy Saturdays vs. rides duration on non-rainy Saturdays), we can use the independent two-sample t-test and we will assign the significance level to 0.5, assuming equal variances from the tests done above.

Null Hypothesis (H0):
The average duration of rides from the Loop to O'Hare International Airport is the same on rainy Saturdays as on non-rainy Saturdays.

Alternative Hypothesis (H1):
The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays.

In [26]:
# Assigning alpha to 0.05
alpha = 0.05

# T-statictic and p-value
t_statistics, p_value = stats.ttest_ind(no_rain_saturdays, rain_saturdays, equal_var=True)

# Print t-statistic and p-value
print("T-statistic:", t_statistics)
print("P-value:", p_value)

# Perform hypothesis testing
if p_value < alpha:
    print("Reject the null hypothesis. The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays.")
else:
    print("Fail to reject the null hypothesis. The average duration of rides from the Loop to O'Hare International Airport is the same on rainy Saturdays as on non-rainy Saturdays.")
    

T-statistic: -6.946177714041499
P-value: 6.517970327099473e-12
Reject the null hypothesis. The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays.


In [27]:
#Being conservative and performing the test with equal_var=False as well since the Levene test was pretty close to 0.5

# Assigning alpha to 0.05
alpha = 0.05

# T-statictic and p-value
t_statistics, p_value = stats.ttest_ind(no_rain_saturdays, rain_saturdays, equal_var=False)

# Print t-statistic and p-value
print("T-statistic:", t_statistics)
print("P-value:", p_value)

# Perform hypothesis testing
if p_value < alpha:
    print("Reject the null hypothesis. The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays.")
else:
    print("Fail to reject the null hypothesis. The average duration of rides from the Loop to O'Hare International Airport is the same on rainy Saturdays as on non-rainy Saturdays.")
    

T-statistic: -7.186034288068629
P-value: 6.738994326108734e-12
Reject the null hypothesis. The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays.


# Final Conclusions

As stated in the section above, **we reject the null hypothesis**. The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays. When comparing the average durations in minutes, it's clear that Rainy Saturdays have longer rides, lasting approximately 40.45 minutes compared to 33.32 minutes on Non-Rainy Saturdays, while the average ride takes approximately 34.52 minutes.


**The Hypotheses Tested:** <br>
Null Hypothesis (H0): The average duration of rides from the Loop to O'Hare International Airport is the same on rainy Saturdays as on non-rainy Saturdays.

Alternative Hypothesis (HA): The average duration of rides from the Loop to O'Hare International Airport differs on rainy Saturdays compared to non-rainy Saturdays.



**Test Summary**<br>
By conducting Levene's test for equality of variances, I assessed whether the variances of the two groups (no_rain_saturdays and rain_saturdays) are significantly different. The p-value from Levene's test was greater than the significance level of 0.05 (p = 0.53), indicating no significant difference in variances between the groups. Therefore, I assumed equal variances and conducted a two-sample t-test with equal_var=True.

The t-test resulted in the following:
T-statistic: -6.946177714041499
P-value: 6.517970327099473e-12

The results led to the rejection of the null hypothesis, indicating that the average duration of rides from the Loop to O'Hare International Airport differs significantly between rainy Saturdays and non-rainy Saturdays.

Although the Levene test did not show a substantial difference in variances, I opted on the cautious side due to the small sample size (only Saturdays from November 2017). To ensure robustness, I also performed a Welch's t-test, assuming equal_var=False with the same significance level of 0.05.

The results from the Welch's test were as follows:
T-statistic: -7.186034288068629
P-value: 6.738994326108734e-12

Once again, the null hypothesis was rejected, confirming that the average duration of rides from the Loop to O'Hare International Airport differs significantly between rainy Saturdays and non-rainy Saturdays.



**Limitations**<br>
Although both tests supported the rejection of the null hypothesis, there are limitations to consider. One major limitation is the small sample size, limited to only Saturdays in November. This constraint may introduce biases or inaccuracies, as factors such as specific construction projects or Thanksgiving travel patterns could skew the data. To obtain a more accurate assessment of the significance of the observed differences, a larger sample size spanning various dates throughout the year, with consideration for different weather conditions, would be necessary.

Overall, while the results provide insights into the differences in ride durations based on weather conditions, caution is warranted in interpreting the findings due to the limitations of the dataset.